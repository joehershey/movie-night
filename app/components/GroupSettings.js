import React, { useState } from "react";
import { View, Text, TouchableWithoutFeedback, TextInput } from "react-native";

import { COLORS, STYLES } from "../assets/saved";
import { FontAwesome5 } from "@expo/vector-icons";

import fetch from "cross-fetch";

function GroupSettings(props) {
  const [edit, setEdit] = useState(false);

  function newCode() {
    //PATCH /group/{group_id}/code
    console.log(props.group_id);
    console.log(props.group_id);

    fetch(props.url + "group/" + props.group_id + "/code", {
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json; charset=utf-8",
        Authorization: "Bearer " + props.token,
      },
      method: "PATCH",
    })
      .then((response) => response.json())
      .then((responseJson) => {
        console.log(responseJson);
        props.setCode();
      })
      .catch((error) => {
        console.error(error);
      });
  }

  return (
    <View
      style={{
        borderColor: COLORS.secondary,
        borderWidth: 3,
        margin: 10,
        borderRadius: 20,
        backgroundColor: "white",
      }}
    >
      <View style={{ alignItems: "center" }}>
        <View
          style={{
            margin: 20,
            flexDirection: "column",
          }}
        >
          <Text style={{ fontSize: 35, fontWeight: "bold" }}>
            Group Settings
          </Text>
        </View>
      </View>

      <View
        style={{
          marginLeft: 15,
          marginRight: 15,
          flexDirection: "row",
          alignItems: "center",
        }}
      >
        <Text
          style={{
            flex: 1,
            margin: 5,
            fontSize: 18,
            fontStyle: "italic",
          }}
        >
          Access Code:
        </Text>
        <View style={{ flex: 1.5, flexDirection: "row" }}>
          <View
            style={{
              flex: 12,
              flexDirection: "row",
              justifyContent: "flex-end",
            }}
          >
            <TouchableWithoutFeedback onPress={() => newCode()}>
              <View
                style={{
                  margin: 10,
                  marginRight: 0,
                  padding: 5,
                  backgroundColor: COLORS.secondary,
                  borderRadius: 10,
                }}
              >
                <FontAwesome5 name="exchange-alt" color="white" size={20} />
              </View>
            </TouchableWithoutFeedback>
            <Text
              style={{
                margin: 10,
                marginRight: 0,
                fontSize: 20,
                fontWeight: "bold",
                textAlign: "right",
              }}
            >
              {props.group_code + " "}
            </Text>
          </View>
        </View>
      </View>
      <View
        style={{
          marginLeft: 15,
          marginRight: 15,
          flexDirection: "row",
          alignItems: "center",
        }}
      >
        <Text
          style={{
            flex: 1,
            margin: 5,
            fontSize: 18,
            fontStyle: "italic",
          }}
        >
          Group Name:
        </Text>
        <View style={{ flex: 1.5, flexDirection: "row" }}>
          <View style={{ flex: 12 }}>
            {!edit && (
              <Text
                style={{
                  fontSize: 20,
                  fontWeight: "bold",
                  textAlign: "right",
                }}
              >
                {props.groupName}
              </Text>
            )}

            {edit && (
              <TextInput
                style={[
                  STYLES.input,
                  STYLES.settingsInput,
                  {
                    borderColor:
                      props.groupName.length < 1 ? COLORS.danger : "lightgrey",
                  },
                ]}
                onChangeText={props.setGroupName}
                textAlign="center"
                value={props.groupName}
                placeholder="New group name"
              />
            )}
          </View>
        </View>
      </View>

      <View
        style={{
          marginLeft: 15,
          marginRight: 15,
          marginBottom: 10,
          flexDirection: "row",
          alignItems: "center",
        }}
      >
        <Text
          style={{
            flex: 1.5,
            margin: 5,
            fontSize: 18,
            fontStyle: "italic",
          }}
        >
          Max. Movies Per User:
        </Text>
        <View style={{ flex: 1, flexDirection: "row" }}>
          <View style={{ flex: 12 }}>
            {!edit && (
              <Text
                style={{
                  fontSize: 20,
                  fontWeight: "bold",
                  textAlign: "right",
                }}
              >
                {props.maxMovies}
              </Text>
            )}

            {edit && (
              <TextInput
                style={[
                  STYLES.input,
                  STYLES.settingsInput,
                  {
                    borderColor:
                      props.maxMovies.length < 1 ? COLORS.danger : "lightgrey",
                  },
                ]}
                onChangeText={props.setMaxMovies}
                value={props.maxMovies}
                textAlign="center"
                keyboardType="number-pad"
                placeholder="Max. # per user"
              />
            )}
          </View>
        </View>
      </View>
      {props.isAdmin && (
        <TouchableWithoutFeedback
          onPress={() => {
            if (edit) {
              setEdit(!edit);
              props.changeGroupSettingsAPI();
            } else {
              setEdit(!edit);
            }
          }}
        >
          <View
            style={{
              alignItems: "center",
              alignSelf: "center",
              justifyContent: "center",
              backgroundColor:
                props.maxMovies.length < 1 || props.groupName.length < 1
                  ? "lightgrey"
                  : COLORS.secondary,
              borderRadius: 10,
              width: "20%",
              paddingLeft: 10,
              paddingRight: 10,
              margin: 10,
              height: 40,
            }}
          >
            <Text style={{ color: "white", margin: 5, fontSize: 18 }}>
              {!edit ? "Edit" : "Save"}
            </Text>
          </View>
        </TouchableWithoutFeedback>
      )}
    </View>
  );
}

export default GroupSettings;
