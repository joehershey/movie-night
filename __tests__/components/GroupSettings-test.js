import GroupSettings from "../../app/components/GroupSettings";
import React from "react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { TouchableWithoutFeedback } from "react-native";
import { shallow, configure } from "enzyme";

configure({ adapter: new Adapter() });

describe("Test Group Settings", () => {
  beforeEach(async () => {
    global.tree = renderer.create(<GroupSettings />);
  });

  it("Group settings renders properly", async () => {
    await expect(global.tree).toMatchSnapshot();
  });

  it("Check for newCode() event and simulate press", () => {
    console.error = jest.fn(); // mutes expected console error of invalid prop
    props = {
      isAdmin: true,
      maxMovies: {
        length: 0,
      },
      groupName: {
        length: 0,
      },
    };
    const container = shallow(<GroupSettings {...props} />);
    const onPress = container.find(TouchableWithoutFeedback).first(); // finds onPress function that calls newCode()
    onPress.props().onPress(); // simulate newCode()
    expect(onPress.exists());
  });

  it("Check for edit button and simulate press", () => {
    props = {
      isAdmin: true,
      maxMovies: {
        length: 0,
      },
      groupName: {
        length: 0,
      },
    };
    const container = shallow(<GroupSettings {...props} />);
    const edit = container.find(TouchableWithoutFeedback).at(1); // finds onPress function that calls setEdit
    edit.props().onPress(); // simulate onPress event
    expect(edit.exists());
  });
});
