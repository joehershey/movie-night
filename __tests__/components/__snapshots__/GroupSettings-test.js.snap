// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Group Settings Group settings renders properly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderColor": "#166fdb",
      "borderRadius": 20,
      "borderWidth": 3,
      "margin": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "column",
          "margin": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 35,
            "fontWeight": "bold",
          }
        }
      >
        Group Settings
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginLeft": 15,
        "marginRight": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "fontStyle": "italic",
          "margin": 5,
        }
      }
    >
      Access Code:
    </Text>
    <View
      style={
        Object {
          "flex": 1.5,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginLeft": 15,
        "marginRight": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "flex": 1,
          "fontSize": 18,
          "fontStyle": "italic",
          "margin": 5,
        }
      }
    >
      Group Name:
    </Text>
    <View
      style={
        Object {
          "flex": 1.5,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 10,
        "marginLeft": 15,
        "marginRight": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "flex": 1.5,
          "fontSize": 18,
          "fontStyle": "italic",
          "margin": 5,
        }
      }
    >
      Max. Movies Per User:
    </Text>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
